shell/platform/windows/windows_lifecycle_manager.cc

```diff
@@ -0,0 +1,126 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "windows_lifecycle_manager.h"
+
+#include <TlHelp32.h>
+#include <WinUser.h>
+#include <Windows.h>
+#include <tchar.h>
+
+#include "flutter/shell/platform/windows/flutter_windows_engine.h"
+
+namespace flutter {
+
+WindowsLifecycleManager::WindowsLifecycleManager(FlutterWindowsEngine* engine)
+    : engine_(engine) {}
+
+WindowsLifecycleManager::~WindowsLifecycleManager() {}
+
+void WindowsLifecycleManager::Quit(UINT exit_code) const {
+  ::PostQuitMessage(exit_code);
+}
+
+bool WindowsLifecycleManager::WindowProc(HWND hwnd,
+                                         UINT msg,
+                                         WPARAM wpar,
+                                         LPARAM lpar,
+                                         LRESULT* result) {
+  switch (msg) {
+    case WM_CLOSE:
+      if (IsLastWindowOfProcess()) {
+        engine_->RequestApplicationQuit(ExitType::cancelable,
+                                       wpar);
+      }
+      return true;
+  }
+  return false;
+}
+
+static BOOL CALLBACK WindowEnumCallback(HWND hwnd, LPARAM user_data) {
+  HWND parent = GetParent(hwnd);
+  if (parent == NULL) {
+    int64_t& count = *reinterpret_cast<int64_t*>(user_data);
+    count++;
+  }
+  return true;
+}
+
+class ThreadSnapshot {
+ public:
+  ThreadSnapshot() {
+    thread_snapshot_ = CreateToolhelp32Snapshot(TH32CS_SNAPTHREAD, 0);
+  }
+  ~ThreadSnapshot() {
+    if (thread_snapshot_ != INVALID_HANDLE_VALUE) {
+      CloseHandle(thread_snapshot_);
+    }
+  }
+
+  std::optional<THREADENTRY32> GetFirstThread() {
```

Comment:
Updated the suggestion a bit later but another way to do this that would avoid the need for the get first, get next pair (and the optional) would be to have this return a std::vector<THREADENTRY32> of all the threads. 

What you've got is great -- I leave it entirely up to you as to which you think is the more readable option.

Reply:
I also slightly prefer returning a `std::vector` but agree both options are good :)